core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala

```diff
@@ -0,0 +1,259 @@
+      offsetToFetch
+    }
+  }
+
+  def handlePartitionsWithErrors(partitions: Iterable[TopicPartition]) {
+    if (partitions.nonEmpty)
+      delayPartitions(partitions, brokerConfig.replicaFetchBackoffMs.toLong)
+  }
+
+  def buildLeaderEpochRequest(allPartitions: Seq[(TopicPartition, PartitionFetchState)]): ResultWithPartitions[Map[TopicPartition, Int]] = {
+    val partitionEpochOpts = allPartitions
+      .filter { case (_, state) => state.isTruncatingLog }
+      .map { case (tp, _) => tp -> epochCacheOpt(tp) }.toMap
+
+    val (partitionsWithEpoch, partitionsWithoutEpoch) = partitionEpochOpts.partition { case (tp, epochCacheOpt) => epochCacheOpt.nonEmpty }
+
+    val result = partitionsWithEpoch.map { case (tp, epochCacheOpt) => tp -> epochCacheOpt.get.latestEpoch() }
+    ResultWithPartitions(result, partitionsWithoutEpoch.keys.toSet)
+  }
+
+  def fetchEpochsFromLeader(partitions: Map[TopicPartition, Int]): Map[TopicPartition, EpochEndOffset] = {
+    partitions.map { case (tp, epoch) =>
+      tp -> new EpochEndOffset(Errors.NONE, replicaMgr.getReplicaOrException(tp).epochs.get.endOffsetFor(epoch))
+    }
+  }
+
+  def maybeTruncate(fetchedEpochs: Map[TopicPartition, EpochEndOffset]): ResultWithPartitions[Map[TopicPartition, Long]] = {
+    val truncationPoints = scala.collection.mutable.HashMap.empty[TopicPartition, Long]
+    val partitionsWithError = mutable.Set[TopicPartition]()
+
+    fetchedEpochs.foreach { case (topicPartition, epochOffset) =>
+      try {
+        val futureReplica = replicaMgr.getReplicaOrException(topicPartition, Request.FutureLocalReplicaId)
+        val partition = replicaMgr.getPartition(topicPartition).get
+
+        if (epochOffset.hasError) {
+          info(s"Retrying leaderEpoch request for partition $topicPartition as the current replica reported an error: ${epochOffset.error}")
+          partitionsWithError += topicPartition
+        } else {
+          val truncationOffset =
+            if (epochOffset.endOffset == UNDEFINED_EPOCH_OFFSET)
+              futureReplica.highWatermark.messageOffset
+            else if (epochOffset.endOffset >= futureReplica.logEndOffset.messageOffset)
+              futureReplica.logEndOffset.messageOffset
+            else
+              epochOffset.endOffset
+
+          partition.truncateTo(truncationOffset, isFuture = true)
+          truncationPoints.put(topicPartition, truncationOffset)
+        }
+      } catch {
+        case e: KafkaStorageException =>
+          info(s"Failed to truncate $topicPartition", e)
+          partitionsWithError += topicPartition
+      }
+    }
+    ResultWithPartitions(truncationPoints, partitionsWithError)
+  }
+
+  def buildFetchRequest(partitionMap: Seq[(TopicPartition, PartitionFetchState)]): ResultWithPartitions[FetchRequest] = {
+    // Only include replica in the fetch request if it is not throttled.
+    val maxPartitionOpt = partitionMap.filter { case (topicPartition, partitionFetchState) =>
```

Comment:
It is called maxPartitionOpt because this is the toipcPartition with largest topic (in alphabetical order) or the largest partition (if topic string is the same). This topic partition is not selected based on its order in partitionStates. `lastPartitionOpt` seems to suggest it is the last partition that is put into the partitionStates. That is why I used `maxPartitionOpt`.

Do you still prefer `lastPartitionOpt`?